nucypher/policy/conditions/wallet.py

```diff
@@ -0,0 +1,143 @@
+from marshmallow import fields, validate
+
+from nucypher.policy.conditions.base import AccessControlCondition
+from nucypher.policy.conditions.context import extract_user_address_from_context
+from nucypher.policy.conditions.exceptions import (
+    InvalidCondition,
+    InvalidConditionContext,
+)
+from nucypher.policy.conditions.lingo import (
+    ConditionType,
+)
+
+
+class WalletAllowlistCondition(AccessControlCondition):
+    """
+    A condition that checks if a user's wallet address is in a list of allowed addresses.
+    The user must provide a signed message to prove ownership of the wallet.
+    """
+
+    CONDITION_TYPE = ConditionType.WALLET_ALLOWLIST.value
+
+    class Schema(AccessControlCondition.Schema):
+        condition_type = fields.Constant(ConditionType.WALLET_ALLOWLIST.value)
+        addresses = fields.List(
+            fields.String(required=True),
+            required=True,
+            validate=validate.Length(min=1, max=25),
+        )
+
+    def __init__(
+        self,
+        addresses: List[str],
+        name: Optional[str] = None,
+    ):
+        """
+        Initialize a WalletAllowlistCondition.
+
+        Args:
+            addresses: List of checksummed wallet addresses that are allowed to decrypt (must be
+             properly checksummed Ethereum addresses or an exception will be raised)
+            name: Optional name for the condition
+        """
+        # Check for duplicates
+        if len(set(addresses)) != len(addresses):
+            raise InvalidCondition("Duplicate addresses are not allowed")
+
+        # Validate all addresses
+        for address in addresses:
+            try:
+                # Ensure addresses have proper checksum
+                normalized_address = to_checksum_address(address)
+            except ValueError as e:
+                raise InvalidCondition(f"Invalid Ethereum address: {address}") from e
+            if normalized_address != address:
+                raise InvalidCondition(
+                    f"Address {address} is not a checksummed address"
+                )
+
+        # Store the validated addresses
+        self.addresses = addresses
+
+        super().__init__(
+            condition_type=self.CONDITION_TYPE,
+            name=name,
+        )
+
+    def verify(self, **context) -> Tuple[bool, Union[None, Dict]]:
+        """
+        Verify if the user's wallet address is in the allowlist.
+
+        Args:
+            context: Dictionary containing context data, including the user's address
+
+        Returns:
+            Tuple containing:
+                - Boolean indicating if verification passed
+                - None (no additional data returned)
+        """
+        if not context:
+            raise InvalidConditionContext(
+                "Context is required for wallet-allowlist condition"
+            )
+
+        # Extract the user's address from the context
+        user_address = extract_user_address_from_context(context)
```

Comment:
Have a look at how this is handled in other conditions: https://github.com/nucypher/nucypher/blob/1dff7d830c2bdb3f35475e2ec5866e71c373d63e/nucypher/policy/conditions/evm.py#L106-L133

This is a generic EVM condition, so let's forget about the RPC execution part, and how `resolved_parameters` is populated. As you can see this implies calling `resolve_any_context_variables`, which automatically handles (1) resolution of `:userAddress` to an ETH address, and, even more importantly, (2) cryptographic validation that the requester owns that address. I'd change your custom context resolution logic to use `resolve_any_context_variables`, similarly as the code above.

Reply:
I updated to use `resolve_any_context_variables` in the commit: https://github.com/nucypher/nucypher/pull/3601/commits/53688d63d16559e6a51bc0c9a290ef3044e5a906
However, what do you think about using `_resolve_user_address` directly after renaming it to `resolve_user_address` or after wrapping it inside a function that provide `USER_ADDRESS_CONTEXT` as the first param?
The new call would look like something similar to:
```py
        user_address = resolve_user_address(**context)["address"] , 
        # or: (and would choose more proper name later)
        user_address = resolve_user_public_address(**context)
```